### PR TITLE
[SMTChecker] Replace constants by their value in-place

### LINKED ISSUE #10631: [SMTChecker] Replace constant variables by their values
Even after https://github.com/ethereum/solidity/pull/10623, we could help the solver even more by replacing the constants directly where they are used, similar to what the code generator does.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/10631

### CODE REVIEW

**Review by axic — DISMISSED** [review]

**Review by hrkrshnn — APPROVED** [review]

**Comment by leonardoalt** [comment]
There's a small bug when the constant is not RationalType, will push fix soon

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -932,7 +932,19 @@ void SMTEncoder::visitObjectCreation(FunctionCall const& _funCall)
 void SMTEncoder::endVisit(Identifier const& _identifier)
 {
 	if (auto decl = identifierToVariable(_identifier))
-		defineExpr(_identifier, currentValue(*decl));
+	{
+		if (decl->isConstant())
+		{
+			RationalNumberType const* rationalType = isConstant(_identifier);
+			solAssert(rationalType, "");
```
Comment (hrkrshnn) [inline libsolidity/formal/SMTEncoder.cpp]: Tests are failing here.

**Review by hrkrshnn — CHANGES_REQUESTED** [review]

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Fixed and updated

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -932,7 +932,26 @@ void SMTEncoder::visitObjectCreation(FunctionCall const& _funCall)
...
+		if (decl->isConstant())
+		{
+			if (RationalNumberType const* rationalType = isConstant(_identifier))
+			{
+				if (rationalType->isNegative())
+					defineExpr(_identifier, smtutil::Expression(u2s(rationalType->literalValue(nullptr))));
+				else
+					defineExpr(_identifier, smtutil::Expression(rationalType->literalValue(nullptr)));
+			}
+			else
```
Comment (axic) [inline libsolidity/formal/SMTEncoder.cpp:945]: Which case is this? Can you add a test for it?

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:945]: There is already a test for it, that's where it broke before this.
`string constant S = "abc";`

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:945]: `test/libsolidity/smtCheckerTests/complex/slither/const_state_variables.sol`

**Review by axic — APPROVED** [review]

**Comment by leonardoalt** [comment]
@axic @chriseth I can't merge because of the OZ tests

**Pull request merged**